ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,437 @@
+require 'some_other_workflow'
+require 'sum_activity'
+require 'some_other_activity'
+
+connection = Temporal::Connection.new('localhost:7233')
+worker = Temporal::Worker.new(
+  connection,
+  namespace: 'demo',
+  task_queue: 'demo',
+  workflows: [SumWorkflow, SomeOtherWorkflow],
+  activities: [SumActivity, SomeOtherActivity]
+)
+
+worker.run
+```
+
+#### Alternative approach
+
+To support passing additional context to the activity classes a worker can be initialised with an
+instance of the activity class (only available for activities).
+
+```ruby
+redis = Redis.new('localhost:6379')
+
+Temporal::Worker.new(
+  ...,
+  activities: [CalculatorActivities.new(redis)]
+)
+```
+
+*NOTE: This however will come with a heavy warning — the activity must be strictly thread-safe as
+the provided instance will be called from multiple threads (and possibly at the same exact time).*
+
+
+## Execute a child workflow
+
+`parent_workflow.rb`
+
+```ruby
+class ChildWorkflow < Temporal::Workflow
+  def execute(name)
+    "Hello, #{name}!"
+  end
+end
+
+class ParentWorkflow < Temporal::Workflow
+  def execute
+    workflow.execute_workflow(ChildWorkflow, 'parent', id: 'id-2')
+  end
+end
+```
+
+
+## Async workflow
+
+Ruby doesn't have a standard (or at least very popular) framework for async execution of the code.
+Because of this we can assume that most Ruby engineers using the SDK won't be immediately familiar
+with writing async code. Because of this we have decided to make async execution opt-in, keeping
+everything synchronous by default.
+
+We have also chosen an eager execution pattern where any `async` block is started immediately and
+performed until blocked before proceeding with the execution. This is a bit easier to get a hold of
+and eliminates the risk of forgetting to actually start an async block. However this means that
+we'll need to find a solution for unhandled promise rejections, which we'll cover separately in a
+more detailed workflow proposal.
+
+We will introduce a helper method `async` to execute a block of code asynchronously:
+
+```ruby
+require 'sum_activity'
+
+class AsyncWorkflow < Temporal::Workflow
+  def execute
+    activity_promise = async { workflow.execute_activity(SumActivity, 1, 2) }
+    workflow_promise = async { workflow.execute_workflow(ChildWorkflow, 'parent') }
+
+    # Block execution until both promises are resolved
+    workflow.wait_for(activity_promise, workflow_promise)
+
+    activity_promise.result # => 3
+    workflow_promise.result # => 'Hello, parent!'
+  end
+end
+```
+
+A bit more involved example:
+
+```ruby
+class AnotherAsyncWorkflow < Temporal::Workflow
+  def execute
+    promise_1 = async { wait_and_sum(1, 2, delay: 10) }
+    promise_2 = async { wait_and_sum(3, 4, delay: 15) }
+
+    # The same exact method can be called in a blocking fashion
+    wait_and_sum(5, 6, delay: 5) # => 11
+
+    # Block execution until any promise is resolved
+    async.wait_for(async.any(promise_1, promise_2))
```

Comment:
In one of our conversations, these combinators returned promises. Doesn't look like this is reflected in the proposal.